### PR TITLE
Migrate P2P references

### PR DESCRIPTION
**Pull request by Sridhar-MS**
Fixes item 2 and 5 in https://github.com/dotnet/cli/issues/4167

- Given a project migrate all the project references recursively.
- Add a boolean option '--skip-project-references' to skip migrating P2P references. By default P2P references are migrated.
- Remove the '--output' since it is not used heavily and makes migrating multiple projects easier.
- Add tests

cc @brthor @eerhardt @livarcocc

### CODE REVIEW

File: src/Microsoft.DotNet.ProjectJsonMigration/ProjectMigrator.cs [hunk 1]
```diff
@@ -33,14 +35,51 @@ public ProjectMigrator(IMigrationRule ruleSet)
...
+
+                var projectContext = migrationRuleInputs.DefaultProjectContext;
+                var projectExports = projectContext.CreateExporter("_").GetDependencies(LibraryType.Project);
+
+                foreach(var export in projectExports)
+                {
+                    var projectFile = ((ProjectDescription)export.Library).Project.ProjectFilePath;
+                    var projectDir = Path.GetDirectoryName(projectFile);
+
+                    if (_migratedProjects.Contains(projectDir))
```
Comment (brthor) [inline src/Microsoft.DotNet.ProjectJsonMigration/ProjectMigrator.cs]: You can have a case where a project is added to the queue multiple times before it's added to _migratedProjects causing duplicate work. 

One solution would be to check _migratedProjects immediately after Dequeueing, and just let the duplicates get on the queue but then ignore them.

File: src/Microsoft.DotNet.ProjectJsonMigration/ProjectMigrator.cs [hunk 2]
```diff
@@ -33,14 +35,51 @@ public ProjectMigrator(IMigrationRule ruleSet)
...
+            if (settings == null)
+            {
+                throw new ArgumentNullException();
+            }
 
-            SetupOutputDirectory(migrationSettings.ProjectDirectory, migrationSettings.OutputDirectory);
+            Queue<MigrationSettings> settingsQueue = new Queue<MigrationSettings>();
+            settingsQueue.Enqueue(settings);
+
+            while(settingsQueue.Count() != 0)
```
Comment (brthor) [inline src/Microsoft.DotNet.ProjectJsonMigration/ProjectMigrator.cs]: What do you think about factoring this into another class or method that returns the full lists of projects to go migrate?

File: src/Microsoft.DotNet.ProjectJsonMigration/ProjectMigrator.cs (hunk 1 above)
Comment (Sridhar-MS) [inline src/Microsoft.DotNet.ProjectJsonMigration/ProjectMigrator.cs]: Good catch. Fixed it

File: src/Microsoft.DotNet.ProjectJsonMigration/ProjectMigrator.cs (hunk 2 above)
Comment (Sridhar-MS) [inline src/Microsoft.DotNet.ProjectJsonMigration/ProjectMigrator.cs]: I am using the project dependency finder now..

File: src/Microsoft.DotNet.ProjectJsonMigration/ProjectDependencyFinder.cs [hunk 3]
```diff
@@ -14,6 +14,17 @@ namespace Microsoft.DotNet.ProjectJsonMigration
 {
     public class ProjectDependencyFinder 
     {
+        public IEnumerable<ProjectDependency> ResolveProjectDependencies(IEnumerable<ProjectContext> projectContexts)
```
Comment (brthor) [inline src/Microsoft.DotNet.ProjectJsonMigration/ProjectDependencyFinder.cs]: please also add the `pre resolved-dependencies` param, this is used so you can avoid unresolved references that are actually csproj references

File: src/Microsoft.DotNet.ProjectJsonMigration/ProjectMigrator.cs [hunk 4]
```diff
@@ -30,9 +33,73 @@ public ProjectMigrator(IMigrationRule ruleSet)
...
 
-        public void Migrate(MigrationSettings migrationSettings)
+        public void Migrate(MigrationSettings rootSettings, bool skipProjectReferences = false)
+        {
+            if (rootSettings == null)
+            {
+                throw new ArgumentNullException();
+            }
+
+            MigrateHelper(rootSettings);
```
Comment (brthor) [inline src/Microsoft.DotNet.ProjectJsonMigration/ProjectMigrator.cs]: can we change the name of this function to reflect it's intention? What does the helper do?

**Comment by brthor** [comment]
A couple small comments, overall looks great. 

After those are addressed :shipit:

File: src/Microsoft.DotNet.ProjectJsonMigration/ProjectDependencyFinder.cs [hunk 5]
```diff
@@ -50,12 +61,12 @@ public IEnumerable<ProjectDependency> ResolveProjectDependencies(ProjectContext
 
         private Dictionary<string, ProjectDependency> FindPossibleProjectDependencies(string projectJsonFilePath)
         {
-            var projectDirectory = Path.GetDirectoryName(projectJsonFilePath);
+            var projectRootDirectory = GetRootFromProjectJson(projectJsonFilePath);
 
             var projectSearchPaths = new List<string>();
-            projectSearchPaths.Add(projectDirectory);
```
Comment (Sridhar-MS) [inline src/Microsoft.DotNet.ProjectJsonMigration/ProjectDependencyFinder.cs]: So @brthor there is no need to add the current directory of the project to search paths?

Comment (brthor) [inline src/Microsoft.DotNet.ProjectJsonMigration/ProjectDependencyFinder.cs]: going to double double check

Comment (brthor) [inline src/Microsoft.DotNet.ProjectJsonMigration/ProjectDependencyFinder.cs]: no we don't need to, looks like nuget doesn't

https://github.com/NuGet/NuGet.Client/blob/125b3bbd41dec25abfae8dd8d64627f78aae41d0/src/NuGet.Core/NuGet.ProjectModel/PackageSpecResolver.cs#L93

File: src/Microsoft.DotNet.ProjectJsonMigration/ProjectMigrator.cs (hunk 4 above)
Comment (Sridhar-MS) [inline src/Microsoft.DotNet.ProjectJsonMigration/ProjectMigrator.cs]: It migrates a single `MigrationSettings`. Any suggestions?

Comment (brthor) [inline src/Microsoft.DotNet.ProjectJsonMigration/ProjectMigrator.cs]: MigrateProjectHelper or just MigrateProject seem okay

File: src/Microsoft.DotNet.ProjectJsonMigration/ProjectDependencyFinder.cs (hunk 3 above)
Comment (Sridhar-MS) [inline src/Microsoft.DotNet.ProjectJsonMigration/ProjectDependencyFinder.cs]: Fixed

**Comment by Sridhar-MS** [comment]
@dotnet-bot test CentOS7.1 x64 Debug Build please

**Comment by brthor** [comment]
After the xproj test :shipit: Looks good

**Comment by Sridhar-MS** [comment]
@dotnet-bot test OpenSUSE13.2 x64 Debug Build please

**Comment by livarcocc** [comment]
@dotnet-bot Test OpenSUSE13.2 x64 Debug Build

**Pull request merged**